src/rgw/rgw_aws_lambda.cc

```diff
@@ -0,0 +1,209 @@
+                 ^ (std::hash<std::string>()(k.accessSecret) << 1)) >> 1)
+               ^ (std::hash<std::string>()(k.arn) << 1)
+               ^ (std::hash<bool>()(k.useSSL) << 1);
+      }
+    };
+
+
+  };
+
+  struct message_t {
+    const Aws::Lambda::LambdaClient *client;
+    const std::string function_name;
+    const std::string payload;
+
+    message_t(const Aws::Lambda::LambdaClient *client, std::string function_name, std::string payload) :
+            client(client),
+            function_name(std::move(function_name)),
+            payload(std::move(payload)) {};
+  };
+
+  typedef std::unordered_map<client_id_t, Aws::Lambda::LambdaClient *, client_id_t::hasher> ClientList;
+
+
+  typedef boost::lockfree::queue<message_t *, boost::lockfree::fixed_sized<true>> MessageQueue;
+
+  class Manager {
+  private:
+    MessageQueue messageQueue;
+    ClientList clientList;
+    std::thread runner;
+    bool stopped;
+    mutable std::mutex clientsLock;
+    CephContext *const cct;
+
+    void publish_internal(message_t *msg) {
+      const std::unique_ptr<message_t> msg_owner(msg);
+      std::cout << "Invoking.." << std::endl;
+      const Aws::Lambda::LambdaClient *client = msg->client;
+      Aws::Lambda::Model::InvokeRequest invokeRequest;
+      invokeRequest.SetInvocationType(Aws::Lambda::Model::InvocationType::RequestResponse);
+      invokeRequest.SetLogType(Aws::Lambda::Model::LogType::Tail);
+      invokeRequest.SetFunctionName(Aws::String(msg->function_name));
+      std::shared_ptr<Aws::IOStream> body = Aws::MakeShared<Aws::StringStream>("Simple Allocation Tag");
+      *body << msg->payload;
+      invokeRequest.SetBody(body);
+      invokeRequest.SetContentType("application/json");
+      auto response = client->Invoke(invokeRequest);
+      if (!response.IsSuccess()) {
+        std::cout << response.GetError().GetMessage() << std::endl;
+      } else {
+        Aws::Lambda::Model::InvokeResult &result = response.GetResult();
+        Aws::IOStream &pl = result.GetPayload();
+        Aws::String returnedData;
+        std::getline(pl, returnedData);
+        std::cout << "Success " << returnedData << "\n" << std::endl;
+      }
+    }
+
+//    Works on messages in the queque
+    void run() {
+      while (!stopped) {
+        const auto count = messageQueue.consume_all(
+                std::bind(&Manager::publish_internal, this, std::placeholders::_1));
+        if (count == 0) {
+          std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+        }
+      }
+    }
+
+  public:
+    Manager(size_t _max_clients, size_t max_queue_size, CephContext *_cct) : messageQueue(max_queue_size),
+                                                                             runner(&Manager::run, this),
+                                                                             stopped(false), cct(_cct),
+                                                                             clientList(_max_clients) {}
+
```

Comment:
manager should also have a dtor, when the client pointers in the client list are deleted